// @confio/relayer/build/module/codec/ibc/core/connection/v1/genesis.js
/* eslint-disable */
import Long from 'long';
import { IdentifiedConnection, ConnectionPaths, } from '../../../../ibc/core/connection/v1/connection';
import _m0 from 'protobufjs/minimal';
export const protobufPackage = 'ibc.core.connection.v1';
const baseGenesisState = { nextConnectionSequence: Long.UZERO };
export const GenesisState = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.connections) {
            IdentifiedConnection.encode(v, writer.uint32(10).fork()).ldelim();
        }
        for (const v of message.clientConnectionPaths) {
            ConnectionPaths.encode(v, writer.uint32(18).fork()).ldelim();
        }
        if (!message.nextConnectionSequence.isZero()) {
            writer.uint32(24).uint64(message.nextConnectionSequence);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof Uint8Array ? new _m0.Reader(input) : input;
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = { ...baseGenesisState };
        message.connections = [];
        message.clientConnectionPaths = [];
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.connections.push(IdentifiedConnection.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.clientConnectionPaths.push(ConnectionPaths.decode(reader, reader.uint32()));
                    break;
                case 3:
                    message.nextConnectionSequence = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = { ...baseGenesisState };
        message.connections = [];
        message.clientConnectionPaths = [];
        if (object.connections !== undefined && object.connections !== null) {
            for (const e of object.connections) {
                message.connections.push(IdentifiedConnection.fromJSON(e));
            }
        }
        if (object.clientConnectionPaths !== undefined &&
            object.clientConnectionPaths !== null) {
            for (const e of object.clientConnectionPaths) {
                message.clientConnectionPaths.push(ConnectionPaths.fromJSON(e));
            }
        }
        if (object.nextConnectionSequence !== undefined &&
            object.nextConnectionSequence !== null) {
            message.nextConnectionSequence = Long.fromString(object.nextConnectionSequence);
        }
        else {
            message.nextConnectionSequence = Long.UZERO;
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        if (message.connections) {
            obj.connections = message.connections.map((e) => e ? IdentifiedConnection.toJSON(e) : undefined);
        }
        else {
            obj.connections = [];
        }
        if (message.clientConnectionPaths) {
            obj.clientConnectionPaths = message.clientConnectionPaths.map((e) => e ? ConnectionPaths.toJSON(e) : undefined);
        }
        else {
            obj.clientConnectionPaths = [];
        }
        message.nextConnectionSequence !== undefined &&
            (obj.nextConnectionSequence = (message.nextConnectionSequence || Long.UZERO).toString());
        return obj;
    },
    fromPartial(object) {
        const message = { ...baseGenesisState };
        message.connections = [];
        message.clientConnectionPaths = [];
        if (object.connections !== undefined && object.connections !== null) {
            for (const e of object.connections) {
                message.connections.push(IdentifiedConnection.fromPartial(e));
            }
        }
        if (object.clientConnectionPaths !== undefined &&
            object.clientConnectionPaths !== null) {
            for (const e of object.clientConnectionPaths) {
                message.clientConnectionPaths.push(ConnectionPaths.fromPartial(e));
            }
        }
        if (object.nextConnectionSequence !== undefined &&
            object.nextConnectionSequence !== null) {
            message.nextConnectionSequence = object.nextConnectionSequence;
        }
        else {
            message.nextConnectionSequence = Long.UZERO;
        }
        return message;
    },
};
