// @confio/relayer/build/main/binary/commander-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addLoggerOptionsTo = exports.chainOption = exports.destOption = exports.srcOption = exports.interactiveOption = exports.mnemonicOption = exports.keyFileOption = exports.homeOption = void 0;
const commander_1 = require("commander");
const create_logger_1 = require("./create-logger");
exports.homeOption = new commander_1.Option('--home <path>', 'home option description (default: $HOME/.ibc-setup)');
exports.keyFileOption = new commander_1.Option('--key-file <path>');
exports.mnemonicOption = new commander_1.Option('--mnemonic <mnemonic>');
exports.interactiveOption = new commander_1.Option('-i, --interactive');
exports.srcOption = new commander_1.Option('--src <chain>');
exports.destOption = new commander_1.Option('--dest <chain>');
exports.chainOption = new commander_1.Option('--chain <chain>');
const addLoggerOptionsTo = (command) => {
    return command
        .addOption(new commander_1.Option('--log-level <level>').choices(Object.keys(create_logger_1.levels)))
        .option('-v, --verbose')
        .option('-q, --quiet')
        .option('--log-file <path>');
};
exports.addLoggerOptionsTo = addLoggerOptionsTo;
