// @confio/relayer/build/main/binary/create-logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLogger = exports.resolveLevel = exports.defaultLevel = exports.levels = void 0;
const winston_1 = __importDefault(require("winston"));
const resolve_option_1 = require("./utils/options/resolve-option");
exports.levels = {
    error: 0,
    warn: 1,
    info: 2,
    verbose: 3,
    debug: 4,
};
exports.defaultLevel = 'info'; // if not provided
function validateLevel(level) {
    return level ? Object.keys(exports.levels).includes(level) : false;
}
function resolveLevel(flags) {
    const level = resolve_option_1.resolveOption('logLevel')(flags.logLevel, process.env.RELAYER_LOG_LEVEL);
    if (level !== null && !validateLevel(level)) {
        return [exports.defaultLevel, level];
    }
    const levelValue = exports.levels[level !== null && level !== void 0 ? level : 'error'];
    if (flags.verbose && levelValue < exports.levels.verbose) {
        return ['verbose', null];
    }
    if (flags.quiet && levelValue <= exports.levels.error) {
        return ['error', null];
    }
    if (level) {
        return [level, null];
    }
    return [exports.defaultLevel, null];
}
exports.resolveLevel = resolveLevel;
function createLogger(flags) {
    const [level, invalidInputLevel] = resolveLevel(flags);
    const fileTransport = flags.logFile
        ? [
            new winston_1.default.transports.File({
                handleExceptions: true,
                filename: flags.logFile,
                format: winston_1.default.format.combine(winston_1.default.format.timestamp()),
            }),
        ]
        : [];
    const logger = winston_1.default.createLogger({
        level,
        levels: exports.levels,
        format: winston_1.default.format.combine(winston_1.default.format.errors({ stack: true }), winston_1.default.format.json()),
        transports: [
            new winston_1.default.transports.Console({
                handleExceptions: true,
                format: winston_1.default.format.combine(winston_1.default.format.colorize(), winston_1.default.format.simple()),
            }),
            ...fileTransport,
        ],
    });
    if (invalidInputLevel !== null) {
        logger.error(`Invalid log-level "${invalidInputLevel}". Please use one of: ${Object.keys(exports.levels)
            .map((level) => `"${level}"`)
            .join(', ')}`);
    }
    return logger;
}
exports.createLogger = createLogger;
