// @confio/relayer/build/main/binary/utils/load-and-validate-app.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadAndValidateApp = void 0;
const fs_1 = __importDefault(require("fs"));
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const ajv_1 = __importDefault(require("ajv"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const constants_1 = require("../constants");
function readAppYaml(filepath) {
    try {
        return fs_1.default.readFileSync(filepath, 'utf-8');
    }
    catch (error) {
        if (error.code !== 'ENOENT') {
            throw error;
        }
        return null;
    }
}
function loadAndValidateApp(home) {
    var _a;
    const appContents = readAppYaml(path_1.default.join(home, constants_1.appFile));
    if (!appContents) {
        return null;
    }
    const app = js_yaml_1.default.load(appContents);
    const ajv = new ajv_1.default({ allErrors: true });
    const schema = {
        type: 'object',
        additionalProperties: false,
        required: [],
        properties: {
            src: { type: 'string', nullable: true, default: null },
            srcConnection: { type: 'string', nullable: true, default: null },
            dest: { type: 'string', nullable: true },
            destConnection: { type: 'string', nullable: true },
            mnemonic: { type: 'string', nullable: true },
            keyFile: { type: 'string', nullable: true },
        },
    };
    const validate = ajv.compile(schema);
    if (!validate(app)) {
        const errors = ((_a = validate.errors) !== null && _a !== void 0 ? _a : []).map(({ dataPath, message }) => `"${dataPath}" ${message}`);
        throw new Error([`${constants_1.appFile} validation failed.`, ...errors].join(os_1.default.EOL));
    }
    return app;
}
exports.loadAndValidateApp = loadAndValidateApp;
