// @confio/relayer/build/main/binary/utils/load-and-validate-registry.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadAndValidateRegistry = void 0;
const fs_1 = __importDefault(require("fs"));
const os_1 = __importDefault(require("os"));
const ajv_1 = __importDefault(require("ajv"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const constants_1 = require("../constants");
function loadAndValidateRegistry(filepath) {
    var _a;
    const registry = js_yaml_1.default.load(fs_1.default.readFileSync(filepath, 'utf-8'));
    const ajv = new ajv_1.default({ allErrors: true });
    const schema = {
        type: 'object',
        required: ['chains', 'version'],
        additionalProperties: false,
        properties: {
            version: {
                type: 'number',
            },
            chains: {
                type: 'object',
                minProperties: 2,
                required: [],
                additionalProperties: false,
                patternProperties: {
                    '^(.*)$': {
                        type: 'object',
                        required: ['chain_id', 'gas_price', 'prefix', 'rpc'],
                        additionalProperties: false,
                        properties: {
                            chain_id: { type: 'string' },
                            prefix: { type: 'string' },
                            gas_price: { type: 'string' },
                            faucet: { type: 'string', nullable: true },
                            hd_path: { type: 'string', nullable: true },
                            ics20_port: { type: 'string', nullable: true },
                            rpc: { type: 'array', items: { type: 'string' }, minItems: 1 },
                        },
                    },
                },
            },
        },
    };
    const validate = ajv.compile(schema);
    if (!validate(registry)) {
        const errors = ((_a = validate.errors) !== null && _a !== void 0 ? _a : []).map(({ dataPath, message }) => `"${dataPath}" ${message}`);
        throw new Error([`${constants_1.registryFile} validation failed.`, ...errors].join(os_1.default.EOL));
    }
    return registry;
}
exports.loadAndValidateRegistry = loadAndValidateRegistry;
