// @confio/relayer/build/lib/endpoint.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Endpoint = void 0;
const encoding_1 = require("@cosmjs/encoding");
const stargate_1 = require("@cosmjs/stargate");
const utils_1 = require("./utils");
/**
 * Endpoint is a wrapper around SigningStargateClient as well as ClientID
 * and ConnectionID. Two Endpoints compose a Link and this should expose all the
 * methods you need to work on one half of an IBC Connection, the higher-level
 * orchestration is handled in Link.
 */
class Endpoint {
    constructor(client, clientID, connectionID) {
        this.client = client;
        this.clientID = clientID;
        this.connectionID = connectionID;
    }
    chainId() {
        return this.client.chainId;
    }
    getLatestCommit() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.client.getCommit();
        });
    }
    getPacketsFromBlockEvents({ minHeight, maxHeight, } = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            let query = `send_packet.packet_connection='${this.connectionID}'`;
            if (minHeight) {
                query = `${query} AND block.height>=${minHeight}`;
            }
            if (maxHeight) {
                query = `${query} AND block.height<=${maxHeight}`;
            }
            const search = yield this.client.tm.blockSearchAll({ query });
            const resultsNested = yield Promise.all(search.blocks.map(({ block }) => __awaiter(this, void 0, void 0, function* () {
                const height = block.header.height;
                const result = yield this.client.tm.blockResults(height);
                return utils_1.parsePacketsFromBlockResult(result).map((packet) => ({
                    packet,
                    height,
                    sender: '',
                }));
            })));
            return [].concat(...resultsNested);
        });
    }
    getPacketsFromTxs({ minHeight, maxHeight, } = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            let query = `send_packet.packet_connection='${this.connectionID}'`;
            if (minHeight) {
                query = `${query} AND tx.height>=${minHeight}`;
            }
            if (maxHeight) {
                query = `${query} AND tx.height<=${maxHeight}`;
            }
            const search = yield this.client.tm.txSearchAll({ query });
            const resultsNested = search.txs.map(({ hash, height, result }) => {
                const parsedLogs = stargate_1.logs.parseRawLog(result.log);
                // we accept message.sender (cosmos-sdk) and message.signer (x/wasm)
                let sender = '';
                try {
                    sender = stargate_1.logs.findAttribute(parsedLogs, 'message', 'sender').value;
                }
                catch (_a) {
                    try {
                        sender = stargate_1.logs.findAttribute(parsedLogs, 'message', 'signer').value;
                    }
                    catch (_b) {
                        this.client.logger.warn(`No message.sender nor message.signer in tx ${encoding_1.toHex(hash)}`);
                    }
                }
                return utils_1.parsePacketsFromLogs(parsedLogs).map((packet) => ({
                    packet,
                    height,
                    sender,
                }));
            });
            return [].concat(...resultsNested);
        });
    }
    // returns all packets (auto-paginates, so be careful about not setting a minHeight)
    querySentPackets({ minHeight, maxHeight, } = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const txsPackets = yield this.getPacketsFromTxs({ minHeight, maxHeight });
            const eventsPackets = yield this.getPacketsFromBlockEvents({
                minHeight,
                maxHeight,
            });
            return []
                .concat(...txsPackets)
                .concat(...eventsPackets);
        });
    }
    // returns all acks (auto-paginates, so be careful about not setting a minHeight)
    queryWrittenAcks({ minHeight, maxHeight, } = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            let query = `write_acknowledgement.packet_connection='${this.connectionID}'`;
            if (minHeight) {
                query = `${query} AND tx.height>=${minHeight}`;
            }
            if (maxHeight) {
                query = `${query} AND tx.height<=${maxHeight}`;
            }
            const search = yield this.client.tm.txSearchAll({ query });
            const resultsNested = search.txs.map(({ height, result }) => {
                const parsedLogs = stargate_1.logs.parseRawLog(result.log);
                // const sender = logs.findAttribute(parsedLogs, 'message', 'sender').value;
                return utils_1.parseAcksFromLogs(parsedLogs).map((ack) => (Object.assign({ height }, ack)));
            });
            return [].concat(...resultsNested);
        });
    }
}
exports.Endpoint = Endpoint;
