// @confio/relayer/build/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.presentPacketData = exports.splitPendingPackets = exports.timeGreater = exports.heightGreater = exports.parseAck = exports.parseAcksFromLogs = exports.parsePacket = exports.parseHeightAttribute = exports.parsePacketsFromLogs = exports.parsePacketsFromBlockResult = exports.buildClientState = exports.buildConsensusState = exports.secondsFromDateNanos = exports.timestampFromDateNanos = exports.mapRpcPubKeyToProto = exports.may = exports.parseRevisionNumber = exports.subtractBlock = exports.ensureIntHeight = exports.toIntHeight = exports.createBroadcastTxErrorMessage = void 0;
const encoding_1 = require("@cosmjs/encoding");
const stargate_1 = require("@cosmjs/stargate");
const { parseEvent } = stargate_1.logs;
const long_1 = __importDefault(require("long"));
const proofs_1 = require("../codec/confio/proofs");
const timestamp_1 = require("../codec/google/protobuf/timestamp");
const channel_1 = require("../codec/ibc/core/channel/v1/channel");
const tendermint_1 = require("../codec/ibc/lightclients/tendermint/v1/tendermint");
function createBroadcastTxErrorMessage(result) {
    return `Error when broadcasting tx ${result.transactionHash} at height ${result.height}. Code: ${result.code}; Raw log: ${result.rawLog}`;
}
exports.createBroadcastTxErrorMessage = createBroadcastTxErrorMessage;
function toIntHeight(height) {
    var _a, _b;
    return (_b = (_a = height === null || height === void 0 ? void 0 : height.revisionHeight) === null || _a === void 0 ? void 0 : _a.toNumber()) !== null && _b !== void 0 ? _b : 0;
}
exports.toIntHeight = toIntHeight;
function ensureIntHeight(height) {
    if (typeof height === 'number') {
        return height;
    }
    return toIntHeight(height);
}
exports.ensureIntHeight = ensureIntHeight;
function subtractBlock(height, count = 1) {
    return {
        revisionNumber: height.revisionNumber,
        revisionHeight: height.revisionHeight.subtract(count),
    };
}
exports.subtractBlock = subtractBlock;
const regexRevNum = new RegExp('-([1-9][0-9]*)$');
function parseRevisionNumber(chainId) {
    const match = chainId.match(regexRevNum);
    if (match && match.length >= 2) {
        return long_1.default.fromString(match[1]);
    }
    return new long_1.default(0);
}
exports.parseRevisionNumber = parseRevisionNumber;
// may will run the transform if value is defined, otherwise returns undefined
function may(transform, value) {
    return value === undefined || value === null ? undefined : transform(value);
}
exports.may = may;
function mapRpcPubKeyToProto(pubkey) {
    if (pubkey === undefined) {
        return undefined;
    }
    if (pubkey.algorithm == 'ed25519') {
        return {
            ed25519: pubkey.data,
            secp256k1: undefined,
        };
    }
    else if (pubkey.algorithm == 'secp256k1') {
        return {
            ed25519: undefined,
            secp256k1: pubkey.data,
        };
    }
    else {
        throw new Error(
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        `Unknown validator pubkey type: ${pubkey.algorithm}`);
    }
}
exports.mapRpcPubKeyToProto = mapRpcPubKeyToProto;
function timestampFromDateNanos(date) {
    var _a;
    const nanos = (date.getTime() % 1000) * 1000000 + ((_a = date.nanoseconds) !== null && _a !== void 0 ? _a : 0);
    return timestamp_1.Timestamp.fromPartial({
        seconds: new long_1.default(date.getTime() / 1000),
        nanos,
    });
}
exports.timestampFromDateNanos = timestampFromDateNanos;
function secondsFromDateNanos(date) {
    return Math.floor(date.getTime() / 1000);
}
exports.secondsFromDateNanos = secondsFromDateNanos;
function buildConsensusState(header) {
    return tendermint_1.ConsensusState.fromPartial({
        timestamp: timestampFromDateNanos(header.time),
        root: {
            hash: header.appHash,
        },
        nextValidatorsHash: header.nextValidatorsHash,
    });
}
exports.buildConsensusState = buildConsensusState;
// Note: we hardcode a number of assumptions, like trust level, clock drift, and assume revisionNumber is 1
function buildClientState(chainId, unbondingPeriodSec, trustPeriodSec, height) {
    // Copied here until https://github.com/confio/ics23/issues/36 is resolved
    // https://github.com/confio/ics23/blob/master/js/src/proofs.ts#L11-L26
    const iavlSpec = {
        leafSpec: {
            prefix: Uint8Array.from([0]),
            hash: proofs_1.HashOp.SHA256,
            prehashValue: proofs_1.HashOp.SHA256,
            prehashKey: proofs_1.HashOp.NO_HASH,
            length: proofs_1.LengthOp.VAR_PROTO,
        },
        innerSpec: {
            childOrder: [0, 1],
            minPrefixLength: 4,
            maxPrefixLength: 12,
            childSize: 33,
            hash: proofs_1.HashOp.SHA256,
        },
    };
    const tendermintSpec = {
        leafSpec: {
            prefix: Uint8Array.from([0]),
            hash: proofs_1.HashOp.SHA256,
            prehashValue: proofs_1.HashOp.SHA256,
            prehashKey: proofs_1.HashOp.NO_HASH,
            length: proofs_1.LengthOp.VAR_PROTO,
        },
        innerSpec: {
            childOrder: [0, 1],
            minPrefixLength: 1,
            maxPrefixLength: 1,
            childSize: 32,
            hash: proofs_1.HashOp.SHA256,
        },
    };
    return tendermint_1.ClientState.fromPartial({
        chainId,
        trustLevel: {
            numerator: long_1.default.fromInt(1),
            denominator: long_1.default.fromInt(3),
        },
        unbondingPeriod: {
            seconds: new long_1.default(unbondingPeriodSec),
        },
        trustingPeriod: {
            seconds: new long_1.default(trustPeriodSec),
        },
        maxClockDrift: {
            seconds: new long_1.default(20),
        },
        latestHeight: height,
        proofSpecs: [iavlSpec, tendermintSpec],
        upgradePath: ['upgrade', 'upgradedIBCState'],
        allowUpdateAfterExpiry: false,
        allowUpdateAfterMisbehaviour: false,
    });
}
exports.buildClientState = buildClientState;
function decodeBase64(value) {
    return Buffer.from(value).toString('binary');
}
function parsePacketsFromBlockResult(result) {
    const allEvents = result.beginBlockEvents
        .concat(...result.endBlockEvents)
        .filter(({ type }) => type === 'send_packet')
        .map(({ type, attributes }) => ({
        type,
        attributes: attributes.map(({ key, value }) => ({
            key: decodeBase64(key),
            value: decodeBase64(value),
        })),
    }));
    const flatEvents = [].concat(allEvents.map(parseEvent));
    return flatEvents.map(parsePacket);
}
exports.parsePacketsFromBlockResult = parsePacketsFromBlockResult;
function parsePacketsFromLogs(logs) {
    // grab all send_packet events from the logs
    const allEvents = logs.map((log) => log.events.filter(({ type }) => type === 'send_packet'));
    const flatEvents = [].concat(...allEvents);
    return flatEvents.map(parsePacket);
}
exports.parsePacketsFromLogs = parsePacketsFromLogs;
function parseHeightAttribute(attribute) {
    var _a;
    const [timeoutRevisionNumber, timeoutRevisionHeight] = (_a = attribute === null || attribute === void 0 ? void 0 : attribute.split('-')) !== null && _a !== void 0 ? _a : [];
    if (!timeoutRevisionHeight || !timeoutRevisionNumber) {
        return undefined;
    }
    const revisionNumber = long_1.default.fromString(timeoutRevisionNumber);
    const revisionHeight = long_1.default.fromString(timeoutRevisionHeight);
    // note: 0 revisionNumber is allowed. If there is bad data, '' or '0-0', we will get 0 for the height
    if (revisionHeight.isZero()) {
        return undefined;
    }
    return { revisionHeight, revisionNumber };
}
exports.parseHeightAttribute = parseHeightAttribute;
function parsePacket({ type, attributes }) {
    if (type !== 'send_packet') {
        throw new Error(`Cannot parse event of type ${type}`);
    }
    const attributesObj = attributes.reduce((acc, { key, value }) => (Object.assign(Object.assign({}, acc), { [key]: value })), {});
    return channel_1.Packet.fromPartial({
        sequence: may(long_1.default.fromString, attributesObj.packet_sequence),
        /** identifies the port on the sending chain. */
        sourcePort: attributesObj.packet_src_port,
        /** identifies the channel end on the sending chain. */
        sourceChannel: attributesObj.packet_src_channel,
        /** identifies the port on the receiving chain. */
        destinationPort: attributesObj.packet_dst_port,
        /** identifies the channel end on the receiving chain. */
        destinationChannel: attributesObj.packet_dst_channel,
        /** actual opaque bytes transferred directly to the application module */
        data: attributesObj.packet_data
            ? encoding_1.toUtf8(attributesObj.packet_data)
            : undefined,
        /** block height after which the packet times out */
        timeoutHeight: parseHeightAttribute(attributesObj.packet_timeout_height),
        /** block timestamp (in nanoseconds) after which the packet times out */
        timeoutTimestamp: may(long_1.default.fromString, attributesObj.packet_timeout_timestamp),
    });
}
exports.parsePacket = parsePacket;
function parseAcksFromLogs(logs) {
    // grab all send_packet events from the logs
    const allEvents = logs.map((log) => log.events.filter(({ type }) => type === 'write_acknowledgement'));
    const flatEvents = [].concat(...allEvents);
    return flatEvents.map(parseAck);
}
exports.parseAcksFromLogs = parseAcksFromLogs;
function parseAck({ type, attributes }) {
    var _a, _b;
    if (type !== 'write_acknowledgement') {
        throw new Error(`Cannot parse event of type ${type}`);
    }
    const attributesObj = attributes.reduce((acc, { key, value }) => (Object.assign(Object.assign({}, acc), { [key]: value })), {});
    const originalPacket = channel_1.Packet.fromPartial({
        sequence: may(long_1.default.fromString, attributesObj.packet_sequence),
        /** identifies the port on the sending chain. */
        sourcePort: attributesObj.packet_src_port,
        /** identifies the channel end on the sending chain. */
        sourceChannel: attributesObj.packet_src_channel,
        /** identifies the port on the receiving chain. */
        destinationPort: attributesObj.packet_dst_port,
        /** identifies the channel end on the receiving chain. */
        destinationChannel: attributesObj.packet_dst_channel,
        /** actual opaque bytes transferred directly to the application module */
        data: encoding_1.toUtf8((_a = attributesObj.packet_data) !== null && _a !== void 0 ? _a : ''),
        /** block height after which the packet times out */
        timeoutHeight: parseHeightAttribute(attributesObj.packet_timeout_height),
        /** block timestamp (in nanoseconds) after which the packet times out */
        timeoutTimestamp: may(long_1.default.fromString, attributesObj.packet_timeout_timestamp),
    });
    const acknowledgement = encoding_1.toUtf8((_b = attributesObj.packet_ack) !== null && _b !== void 0 ? _b : '');
    return {
        acknowledgement,
        originalPacket,
    };
}
exports.parseAck = parseAck;
// return true if a > b, or a undefined
function heightGreater(a, b) {
    if (a === undefined) {
        return true;
    }
    // comparing longs made some weird issues (maybe signed/unsigned)?
    // convert to numbers to compare safely
    const [numA, heightA, numB, heightB] = [
        a.revisionNumber.toNumber(),
        a.revisionHeight.toNumber(),
        b.revisionNumber.toNumber(),
        b.revisionHeight.toNumber(),
    ];
    const valid = numA > numB || (numA == numB && heightA > heightB);
    return valid;
}
exports.heightGreater = heightGreater;
// return true if a > b, or a 0
// note a is nanoseconds, while b is seconds
function timeGreater(a, b) {
    if (a === undefined || a.isZero()) {
        return true;
    }
    const valid = a.toNumber() > b * 1000000000;
    return valid;
}
exports.timeGreater = timeGreater;
// take height and time from receiving chain to see which packets have timed out
// return [toSubmit, toTimeout].
// you can advance height, time a block or two into the future if you wish a margin of error
function splitPendingPackets(currentHeight, currentTime, // in seconds
packets) {
    return packets.reduce((acc, packet) => {
        const validPacket = heightGreater(packet.packet.timeoutHeight, currentHeight) &&
            timeGreater(packet.packet.timeoutTimestamp, currentTime);
        return validPacket
            ? Object.assign(Object.assign({}, acc), { toSubmit: [...acc.toSubmit, packet] }) : Object.assign(Object.assign({}, acc), { toTimeout: [...acc.toTimeout, packet] });
    }, {
        toSubmit: [],
        toTimeout: [],
    });
}
exports.splitPendingPackets = splitPendingPackets;
function presentPacketData(data) {
    try {
        return JSON.parse(encoding_1.fromUtf8(data));
    }
    catch (_a) {
        return { hex: encoding_1.toHex(data) };
    }
}
exports.presentPacketData = presentPacketData;
