// @confio/relayer/build/module/lib/endpoint.js
import { toHex } from '@cosmjs/encoding';
import { logs } from '@cosmjs/launchpad';
import { parseRawLog } from '@cosmjs/stargate';
import { parseAcksFromLogs, parsePacketsFromLogs } from './utils';
/**
 * Endpoint is a wrapper around SigningStargateClient as well as ClientID
 * and ConnectionID. Two Endpoints compose a Link and this should expose all the
 * methods you need to work on one half of an IBC Connection, the higher-level
 * orchestration is handled in Link.
 */
export class Endpoint {
    constructor(client, clientID, connectionID) {
        this.client = client;
        this.clientID = clientID;
        this.connectionID = connectionID;
    }
    chainId() {
        return this.client.chainId;
    }
    async getLatestCommit() {
        return this.client.getCommit();
    }
    // returns all packets (auto-paginates, so be careful about not setting a minHeight)
    async querySentPackets({ minHeight, maxHeight, } = {}) {
        let query = `send_packet.packet_connection='${this.connectionID}'`;
        if (minHeight) {
            query = `${query} AND tx.height>=${minHeight}`;
        }
        if (maxHeight) {
            query = `${query} AND tx.height<=${maxHeight}`;
        }
        const search = await this.client.tm.txSearchAll({ query });
        const resultsNested = search.txs.map(({ hash, height, result }) => {
            const parsedLogs = parseRawLog(result.log);
            // we accept message.sender (cosmos-sdk) and message.signer (x/wasm)
            let sender = '';
            try {
                sender = logs.findAttribute(parsedLogs, 'message', 'sender').value;
            }
            catch {
                try {
                    sender = logs.findAttribute(parsedLogs, 'message', 'signer').value;
                }
                catch {
                    this.client.logger.warn(`No message.sender nor message.signer in tx ${toHex(hash)}`);
                }
            }
            return parsePacketsFromLogs(parsedLogs).map((packet) => ({
                packet,
                height,
                sender,
            }));
        });
        return [].concat(...resultsNested);
    }
    // returns all acks (auto-paginates, so be careful about not setting a minHeight)
    async queryWrittenAcks({ minHeight, maxHeight, } = {}) {
        let query = `write_acknowledgement.packet_connection='${this.connectionID}'`;
        if (minHeight) {
            query = `${query} AND tx.height>=${minHeight}`;
        }
        if (maxHeight) {
            query = `${query} AND tx.height<=${maxHeight}`;
        }
        const search = await this.client.tm.txSearchAll({ query });
        const resultsNested = search.txs.map(({ height, result }) => {
            const parsedLogs = parseRawLog(result.log);
            // const sender = logs.findAttribute(parsedLogs, 'message', 'sender').value;
            return parseAcksFromLogs(parsedLogs).map((ack) => ({
                height,
                ...ack,
            }));
        });
        return [].concat(...resultsNested);
    }
}
