// @confio/relayer/build/module/binary/ibc-setup/commands/connect.js
import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';
import { Link } from '../../../lib/link';
import { appFile, registryFile } from '../../constants';
import { loadAndValidateApp } from '../../utils/load-and-validate-app';
import { loadAndValidateRegistry } from '../../utils/load-and-validate-registry';
import { resolveOption } from '../../utils/options/resolve-option';
import { resolveHomeOption } from '../../utils/options/shared/resolve-home-option';
import { resolveKeyFileOption } from '../../utils/options/shared/resolve-key-file-option';
import { resolveMnemonicOption } from '../../utils/options/shared/resolve-mnemonic-option';
import { signingClient } from '../../utils/signing-client';
export async function connect(flags, logger) {
    const home = resolveHomeOption({ homeFlag: flags.home });
    const app = loadAndValidateApp(home);
    if (!app) {
        throw new Error(`${appFile} not found at ${home}`);
    }
    const keyFile = resolveKeyFileOption({ keyFileFlag: flags.keyFile, app });
    const mnemonic = await resolveMnemonicOption({
        interactiveFlag: flags.interactive,
        mnemonicFlag: flags.mnemonic,
        keyFile: keyFile,
        app,
    });
    const src = resolveOption('src', { required: true })(flags.src, app?.src, process.env.RELAYER_SRC);
    const dest = resolveOption('dest', { required: true })(flags.dest, app?.dest, process.env.RELAYER_DEST);
    const options = {
        home,
        mnemonic,
        src,
        dest,
    };
    await run(options, app, logger);
}
export async function run(options, app, logger) {
    const registryFilePath = path.join(options.home, registryFile);
    const registry = loadAndValidateRegistry(registryFilePath);
    const srcChain = registry.chains[options.src];
    if (!srcChain) {
        throw new Error(`src channel  "${options.src}" not found in registry`);
    }
    const destChain = registry.chains[options.dest];
    if (!destChain) {
        throw new Error(`dest channel  "${options.dest}" not found in registry`);
    }
    const nodeA = await signingClient(srcChain, options.mnemonic);
    const nodeB = await signingClient(destChain, options.mnemonic);
    const link = await Link.createWithNewConnections(nodeA, nodeB);
    const appYaml = yaml.dump({
        ...app,
        srcConnection: link.endA.connectionID,
        destConnection: link.endB.connectionID,
    }, {
        lineWidth: 1000,
    });
    fs.writeFileSync(path.join(options.home, appFile), appYaml, {
        encoding: 'utf-8',
    });
    logger.info(`Created connections ${link.endA.connectionID} (${link.endA.clientID}) <=> ${link.endB.connectionID} (${link.endB.clientID})`);
}
