// @confio/relayer/build/module/binary/ibc-setup/commands/ics20.js
import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';
import { Order } from '../../../codec/ibc/core/channel/v1/channel';
import { Link } from '../../../lib/link';
import { appFile, registryFile } from '../../constants';
import { loadAndValidateApp } from '../../utils/load-and-validate-app';
import { loadAndValidateRegistry } from '../../utils/load-and-validate-registry';
import { resolveOption } from '../../utils/options/resolve-option';
import { resolveHomeOption } from '../../utils/options/shared/resolve-home-option';
import { resolveKeyFileOption } from '../../utils/options/shared/resolve-key-file-option';
import { resolveMnemonicOption } from '../../utils/options/shared/resolve-mnemonic-option';
import { signingClient } from '../../utils/signing-client';
const defaultPort = 'transfer';
function resolveConnections({ srcConnection, destConnection, }) {
    if (!srcConnection && destConnection) {
        throw new Error(`You have defined "destConnection" but no "srcConnection". Both "srcConnection" and "destConnection" must be present.`);
    }
    if (srcConnection && !destConnection) {
        throw new Error(`You have defined "srcConnection" but no "destConnection". Both "srcConnection" and "destConnection" must be present.`);
    }
    if (srcConnection && destConnection) {
        return {
            src: srcConnection,
            dest: destConnection,
        };
    }
    return null;
}
export async function ics20(flags, logger) {
    const home = resolveHomeOption({ homeFlag: flags.home });
    const app = loadAndValidateApp(home);
    if (!app) {
        throw new Error(`${appFile} not found at ${home}`);
    }
    const keyFile = resolveKeyFileOption({ keyFileFlag: flags.keyFile, app });
    const mnemonic = await resolveMnemonicOption({
        interactiveFlag: flags.interactive,
        mnemonicFlag: flags.mnemonic,
        keyFile,
        app,
    });
    const src = resolveOption('src', { required: true })(flags.src, app.src, process.env.RELAYER_SRC);
    const dest = resolveOption('dest', { required: true })(flags.dest, app.dest, process.env.RELAYER_DEST);
    // we apply default ports later, once we have the registry
    const srcPort = resolveOption('srcPort')(flags.srcPort, process.env.RELAYER_SRC_PORT);
    const destPort = resolveOption('destPort')(flags.destPort, process.env.RELAYER_DEST_PORT);
    const connections = resolveConnections(app);
    run({
        src,
        dest,
        home,
        mnemonic,
        srcPort,
        destPort,
        connections,
    }, app, logger);
}
async function resolveLink(nodeA, nodeB, connections, logger) {
    if (connections) {
        const link = await Link.createWithExistingConnections(nodeA, nodeB, connections.src, connections.dest, logger);
        logger.info(`Used existing connections ${link.endA.connectionID} (${link.endA.clientID}) <=> ${link.endB.connectionID} (${link.endB.clientID})`);
        return link;
    }
    const link = await Link.createWithNewConnections(nodeA, nodeB);
    logger.info(`Created connections ${link.endA.connectionID} (${link.endA.clientID}) <=> ${link.endB.connectionID} (${link.endB.clientID})`);
    return link;
}
export async function run(options, app, logger) {
    const registryFilePath = path.join(options.home, registryFile);
    const { chains } = loadAndValidateRegistry(registryFilePath);
    const srcChain = chains[options.src];
    if (!srcChain) {
        throw new Error('src chain not found in registry');
    }
    const destChain = chains[options.dest];
    if (!destChain) {
        throw new Error('dest chain not found in registry');
    }
    const ordering = Order.ORDER_UNORDERED;
    const version = 'ics20-1';
    const nodeA = await signingClient(srcChain, options.mnemonic, logger);
    const nodeB = await signingClient(destChain, options.mnemonic, logger);
    const link = await resolveLink(nodeA, nodeB, options.connections, logger);
    const srcConnection = link.endA.connectionID;
    const destConnection = link.endB.connectionID;
    const appFilePath = path.join(options.home, appFile);
    const appYaml = yaml.dump({
        ...app,
        srcConnection,
        destConnection,
    }, {
        lineWidth: 1000,
    });
    fs.writeFileSync(appFilePath, appYaml, { encoding: 'utf-8' });
    // provide default port, either from registry or global default
    const srcPort = resolveOption('src-port', { required: true })(options.srcPort, srcChain.ics20_port, defaultPort);
    const destPort = resolveOption('dest-port', { required: true })(options.destPort, destChain.ics20_port, defaultPort);
    const channels = await link.createChannel('A', srcPort, destPort, ordering, version);
    logger.info(`Created channels for connections ${link.endA.connectionID} <=> ${link.endB.connectionID}: ${channels.src.channelId} (${channels.src.portId}) => ${channels.dest.channelId} (${channels.dest.portId})`);
}
