// @confio/relayer/build/binary/commander-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.helpOptions = exports.addLoggerOptionsTo = exports.destPort = exports.srcPort = exports.destConnection = exports.srcConnection = exports.destTrust = exports.srcTrust = exports.chainOption = exports.destOption = exports.srcOption = exports.interactiveOption = exports.mnemonicOption = exports.keyFileOption = exports.homeOption = void 0;
const commander_1 = require("commander");
const create_logger_1 = require("./create-logger");
exports.homeOption = new commander_1.Option('--home <path>', "Path to relayer's home (default: $HOME/.ibc-setup)");
const keyFileOption = (type) => new commander_1.Option('--key-file <path>', `Path to file to ${type === 'read' ? 'read mnemonic from' : 'write mnemonic to'}`);
exports.keyFileOption = keyFileOption;
exports.mnemonicOption = new commander_1.Option('--mnemonic <mnemonic>', 'BIP39 mnemonic');
exports.interactiveOption = new commander_1.Option('-i, --interactive', 'Read mnemonic from stdin');
exports.srcOption = new commander_1.Option('--src <chain>', 'Source chain from the registry');
exports.destOption = new commander_1.Option('--dest <chain>', 'Destination chain from the registry');
exports.chainOption = new commander_1.Option('--chain <chain>', 'Chain to run query against');
exports.srcTrust = new commander_1.Option('--src-trust <seconds>', 'Trusting period for source connection');
exports.destTrust = new commander_1.Option('--dest-trust <seconds>', 'Trusting period for destination connection');
exports.srcConnection = new commander_1.Option('--src-connection <connection>', 'Source connection id');
exports.destConnection = new commander_1.Option('--dest-connection <connection>', 'Destination connection id');
const srcPort = (descriptionPostfix = '') => new commander_1.Option('--src-port <port>', `Source port to create channel${descriptionPostfix}`);
exports.srcPort = srcPort;
const destPort = (descriptionPostfix = '') => new commander_1.Option('--dest-port <port>', `Destination port to create channel${descriptionPostfix}`);
exports.destPort = destPort;
const addLoggerOptionsTo = (command) => {
    return command
        .addOption(new commander_1.Option('--log-level <level>', `Set log level (default: ${create_logger_1.defaultLevel})`).choices(Object.keys(create_logger_1.levels)))
        .option('-v, --verbose', 'Alias for "--log-level verbose (default: false)"')
        .option('-q, --quiet', 'Alias for "--log-level error (default: false)"')
        .option('--log-file <path>', 'Path to file to write logs to')
        .option('--stack-trace', 'Print stack trace along with log messages (default: false)');
};
exports.addLoggerOptionsTo = addLoggerOptionsTo;
exports.helpOptions = ['-h, --help', 'Display help command'];
