// @confio/relayer/build/binary/create-logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLogger = exports.resolveLevel = exports.defaultLevel = exports.levels = void 0;
const fast_safe_stringify_1 = __importDefault(require("fast-safe-stringify"));
const triple_beam_1 = require("triple-beam");
const winston_1 = __importDefault(require("winston"));
const resolve_option_1 = require("./utils/options/resolve-option");
exports.levels = {
    error: 0,
    warn: 1,
    info: 2,
    verbose: 3,
    debug: 4,
};
exports.defaultLevel = 'info'; // if not provided
function validateLevel(level) {
    return level ? Object.keys(exports.levels).includes(level) : false;
}
function resolveLevel(flags) {
    const level = resolve_option_1.resolveOption('logLevel')(flags.logLevel, process.env.RELAYER_LOG_LEVEL);
    if (level !== null && !validateLevel(level)) {
        return [exports.defaultLevel, level];
    }
    const levelValue = exports.levels[level !== null && level !== void 0 ? level : 'error'];
    if (flags.verbose && levelValue < exports.levels.verbose) {
        return ['verbose', null];
    }
    if (flags.quiet && levelValue <= exports.levels.error) {
        return ['error', null];
    }
    if (level) {
        return [level, null];
    }
    return [exports.defaultLevel, null];
}
exports.resolveLevel = resolveLevel;
function createLogger(flags) {
    const [level, invalidInputLevel] = resolveLevel(flags);
    const fileTransport = flags.logFile
        ? [
            new winston_1.default.transports.File({
                handleExceptions: true,
                filename: flags.logFile,
                format: winston_1.default.format.combine(winston_1.default.format.timestamp()),
            }),
        ]
        : [];
    const logger = winston_1.default.createLogger({
        level,
        levels: exports.levels,
        format: winston_1.default.format.combine(winston_1.default.format.errors({ stack: true }), winston_1.default.format.json()),
        transports: [
            new winston_1.default.transports.Console({
                handleExceptions: true,
                format: winston_1.default.format.combine(winston_1.default.format.colorize(), simpleFormat(flags.stackTrace)),
            }),
            ...fileTransport,
        ],
    });
    if (invalidInputLevel !== null) {
        logger.error(`Invalid log-level "${invalidInputLevel}". Please use one of: ${Object.keys(exports.levels)
            .map((level) => `"${level}"`)
            .join(', ')}`);
    }
    return logger;
}
exports.createLogger = createLogger;
// Heavily based on https://github.com/winstonjs/logform/blob/master/simple.js
function simpleFormat(stackTrace) {
    return winston_1.default.format((info) => {
        let stringifiedRest = fast_safe_stringify_1.default(Object.assign(Object.assign(Object.assign({}, info), { level: undefined, message: undefined, label: undefined }), (stackTrace ? {} : { stack: undefined })));
        stringifiedRest = stringifiedRest !== '{}' ? ` ${stringifiedRest}` : '';
        const label = info.label ? ` [${info.label}]` : '';
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore: indexing with symbols: https://github.com/microsoft/TypeScript/issues/1863
        info[triple_beam_1.MESSAGE] = `${info.level}:${label} ${info.message}${stringifiedRest}`;
        return info;
    })();
}
