// @confio/relayer/build/main/binary/ibc-setup/commands/balances.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = exports.balances = void 0;
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const crypto_1 = require("@cosmjs/crypto");
const launchpad_1 = require("@cosmjs/launchpad");
const proto_signing_1 = require("@cosmjs/proto-signing");
const ibcclient_1 = require("../../../lib/ibcclient");
const constants_1 = require("../../constants");
const load_and_validate_app_1 = require("../../utils/load-and-validate-app");
const load_and_validate_registry_1 = require("../../utils/load-and-validate-registry");
const resolve_home_option_1 = require("../../utils/options/shared/resolve-home-option");
const resolve_key_file_option_1 = require("../../utils/options/shared/resolve-key-file-option");
const resolve_mnemonic_option_1 = require("../../utils/options/shared/resolve-mnemonic-option");
const keys_list_1 = require("./keys-list");
async function balances(flags, logger) {
    const home = resolve_home_option_1.resolveHomeOption({ homeFlag: flags.home });
    const app = load_and_validate_app_1.loadAndValidateApp(home);
    const keyFile = resolve_key_file_option_1.resolveKeyFileOption({ keyFileFlag: flags.keyFile, app });
    const mnemonic = await resolve_mnemonic_option_1.resolveMnemonicOption({
        interactiveFlag: flags.interactive,
        mnemonicFlag: flags.mnemonic,
        keyFile: keyFile,
        app,
    });
    const options = {
        home,
        mnemonic,
    };
    await run(options, logger);
}
exports.balances = balances;
async function run(options, logger) {
    const registryFilePath = path_1.default.join(options.home, constants_1.registryFile);
    const registry = load_and_validate_registry_1.loadAndValidateRegistry(registryFilePath);
    const addresses = await keys_list_1.getAddresses(registry.chains, options.mnemonic);
    const balances = (await Promise.all(addresses.map(async ([chain, data, address]) => {
        const signer = await proto_signing_1.DirectSecp256k1HdWallet.fromMnemonic(options.mnemonic, data.hd_path ? crypto_1.stringToPath(data.hd_path) : undefined, data.prefix);
        const gasPrice = launchpad_1.GasPrice.fromString(data.gas_price);
        const client = await ibcclient_1.IbcClient.connectWithSigner(data.rpc[0], // rpc[0] is guaranteed to be defined by registry validator
        signer, address, {
            prefix: data.prefix,
            gasPrice,
        });
        const coin = await client.query.bank.unverified.balance(address, gasPrice.denom);
        return [chain, coin];
    })))
        .filter(([, coin]) => coin.amount !== '0')
        .map(([chain, coin]) => `${chain}: ${coin.amount}${coin.denom}`)
        .join(os_1.default.EOL);
    if (!balances) {
        logger.info('No funds found for default denomination on any chain.');
        return;
    }
    logger.info(balances);
}
exports.run = run;
