// @confio/relayer/build/main/binary/ibc-setup/commands/init.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = exports.init = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const faucet_client_1 = require("@cosmjs/faucet-client");
const axios_1 = __importDefault(require("axios"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const constants_1 = require("../../constants");
const types_1 = require("../../types");
const derive_address_1 = require("../../utils/derive-address");
const generate_mnemonic_1 = require("../../utils/generate-mnemonic");
const get_default_home_path_1 = require("../../utils/get-default-home-path");
const load_and_validate_registry_1 = require("../../utils/load-and-validate-registry");
const resolve_option_1 = require("../../utils/options/resolve-option");
async function init(flags, logger) {
    const options = {
        src: resolve_option_1.resolveOption('src', {
            required: true,
        })(flags.src, process.env.RELAYER_SRC),
        dest: resolve_option_1.resolveOption('dest', {
            required: true,
        })(flags.dest, process.env.RELAYER_DEST),
        home: resolve_option_1.resolveOption('home', { required: true })(flags.home, process.env.RELAYER_HOME, get_default_home_path_1.getDefaultHomePath),
    };
    await run(options, logger);
}
exports.init = init;
async function run(options, logger) {
    const appFilePath = path_1.default.join(options.home, constants_1.appFile);
    if (fs_1.default.existsSync(appFilePath)) {
        logger.info(`The ${constants_1.appFile} is already initialized at ${options.home}`);
        return;
    }
    if (!fs_1.default.existsSync(options.home)) {
        fs_1.default.mkdirSync(options.home, { recursive: true });
        logger.info(`Initialized home directory at ${options.home}`);
    }
    else if (!fs_1.default.lstatSync(options.home).isDirectory()) {
        throw new Error(`${options.home} must be a directory.`);
    }
    const registryFilePath = path_1.default.join(options.home, constants_1.registryFile);
    if (!fs_1.default.existsSync(registryFilePath)) {
        try {
            const registryFromRemote = await axios_1.default.get('https://raw.githubusercontent.com/confio/ts-relayer/main/demo/registry.yaml');
            fs_1.default.writeFileSync(registryFilePath, registryFromRemote.data);
        }
        catch (error) {
            throw new Error(`Cannot fetch ${constants_1.registryFile} from remote. ${error}`);
        }
    }
    else if (!fs_1.default.lstatSync(registryFilePath).isFile()) {
        throw new Error(`${registryFilePath} must be a file.`);
    }
    const registry = load_and_validate_registry_1.loadAndValidateRegistry(registryFilePath);
    const [chainSrc, chainDest] = [options.src, options.dest].map((chain) => {
        const chainData = registry.chains[chain];
        if (!chainData) {
            throw new Error(`Chain ${chain} is missing in the registry, either check the spelling or add the chain definition to ${registryFilePath}`);
        }
        return chainData;
    });
    const mnemonic = generate_mnemonic_1.generateMnemonic();
    const appYaml = js_yaml_1.default.dump({
        src: options.src,
        dest: options.dest,
        mnemonic,
    }, {
        lineWidth: 1000,
    });
    fs_1.default.writeFileSync(appFilePath, appYaml, { encoding: 'utf-8' });
    logger.info(`Saved configuration to ${appFilePath}`);
    const [addressSrc, addressDest] = await Promise.all([
        derive_address_1.deriveAddress(mnemonic, chainSrc.prefix, chainSrc.hd_path),
        derive_address_1.deriveAddress(mnemonic, chainDest.prefix, chainDest.hd_path),
    ]);
    logger.info(`Source address: ${addressSrc}`);
    logger.info(`Destination address: ${addressDest}`);
    // if there are faucets, ask for tokens
    if (chainSrc.faucet) {
        const srcDenom = types_1.feeDenom(chainSrc);
        logger.info(`Requesting ${srcDenom} for ${chainSrc.chain_id}...`);
        await new faucet_client_1.FaucetClient(chainSrc.faucet).credit(addressSrc, srcDenom);
    }
    if (chainDest.faucet) {
        const destDenom = types_1.feeDenom(chainDest);
        logger.info(`Requesting ${destDenom} for ${chainDest.chain_id}...`);
        await new faucet_client_1.FaucetClient(chainDest.faucet).credit(addressDest, destDenom);
    }
}
exports.run = run;
