// @confio/relayer/build/main/binary/ibc-setup/commands/keys-list.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = exports.getAddresses = exports.keysList = void 0;
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const constants_1 = require("../../constants");
const derive_address_1 = require("../../utils/derive-address");
const load_and_validate_app_1 = require("../../utils/load-and-validate-app");
const load_and_validate_registry_1 = require("../../utils/load-and-validate-registry");
const resolve_home_option_1 = require("../../utils/options/shared/resolve-home-option");
const resolve_key_file_option_1 = require("../../utils/options/shared/resolve-key-file-option");
const resolve_mnemonic_option_1 = require("../../utils/options/shared/resolve-mnemonic-option");
async function keysList(flags, logger) {
    const home = resolve_home_option_1.resolveHomeOption({ homeFlag: flags.home });
    const app = load_and_validate_app_1.loadAndValidateApp(home);
    const keyFile = resolve_key_file_option_1.resolveKeyFileOption({ keyFileFlag: flags.keyFile, app });
    const mnemonic = await resolve_mnemonic_option_1.resolveMnemonicOption({
        interactiveFlag: flags.interactive,
        mnemonicFlag: flags.mnemonic,
        keyFile: keyFile,
        app,
    });
    const options = {
        home,
        mnemonic,
    };
    await run(options, logger);
}
exports.keysList = keysList;
async function getAddresses(registryChains, mnemonic) {
    const chains = Object.entries(registryChains);
    return (await Promise.all(chains.map(([, data]) => derive_address_1.deriveAddress(mnemonic, data.prefix, data.hd_path)))).map((address, index) => [chains[index][0], chains[index][1], address]);
}
exports.getAddresses = getAddresses;
async function run(options, logger) {
    const registryFilePath = path_1.default.join(options.home, constants_1.registryFile);
    const registry = load_and_validate_registry_1.loadAndValidateRegistry(registryFilePath);
    const addresses = (await getAddresses(registry.chains, options.mnemonic))
        .map(([chain, , address]) => `${chain}: ${address}`)
        .join(os_1.default.EOL);
    logger.info(addresses);
}
exports.run = run;
