// @confio/relayer/build/module/codec/tendermint/libs/bits/types.js
/* eslint-disable */
import Long from 'long';
import _m0 from 'protobufjs/minimal';
export const protobufPackage = 'tendermint.libs.bits';
const baseBitArray = { bits: Long.ZERO, elems: Long.UZERO };
export const BitArray = {
    encode(message, writer = _m0.Writer.create()) {
        if (!message.bits.isZero()) {
            writer.uint32(8).int64(message.bits);
        }
        writer.uint32(18).fork();
        for (const v of message.elems) {
            writer.uint64(v);
        }
        writer.ldelim();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof Uint8Array ? new _m0.Reader(input) : input;
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = { ...baseBitArray };
        message.elems = [];
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.bits = reader.int64();
                    break;
                case 2:
                    if ((tag & 7) === 2) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.elems.push(reader.uint64());
                        }
                    }
                    else {
                        message.elems.push(reader.uint64());
                    }
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = { ...baseBitArray };
        message.elems = [];
        if (object.bits !== undefined && object.bits !== null) {
            message.bits = Long.fromString(object.bits);
        }
        else {
            message.bits = Long.ZERO;
        }
        if (object.elems !== undefined && object.elems !== null) {
            for (const e of object.elems) {
                message.elems.push(Long.fromString(e));
            }
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.bits !== undefined &&
            (obj.bits = (message.bits || Long.ZERO).toString());
        if (message.elems) {
            obj.elems = message.elems.map((e) => (e || Long.UZERO).toString());
        }
        else {
            obj.elems = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = { ...baseBitArray };
        message.elems = [];
        if (object.bits !== undefined && object.bits !== null) {
            message.bits = object.bits;
        }
        else {
            message.bits = Long.ZERO;
        }
        if (object.elems !== undefined && object.elems !== null) {
            for (const e of object.elems) {
                message.elems.push(e);
            }
        }
        return message;
    },
};
