// @confio/relayer/build/binary/ibc-setup/commands/channel.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = exports.channel = exports.defaults = void 0;
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const channel_1 = require("../../../codec/ibc/core/channel/v1/channel");
const link_1 = require("../../../lib/link");
const constants_1 = require("../../constants");
const indent_1 = require("../../utils/indent");
const load_and_validate_app_1 = require("../../utils/load-and-validate-app");
const load_and_validate_registry_1 = require("../../utils/load-and-validate-registry");
const resolve_option_1 = require("../../utils/options/resolve-option");
const resolve_home_option_1 = require("../../utils/options/shared/resolve-home-option");
const resolve_key_file_option_1 = require("../../utils/options/shared/resolve-key-file-option");
const resolve_mnemonic_option_1 = require("../../utils/options/shared/resolve-mnemonic-option");
const signing_client_1 = require("../../utils/signing-client");
exports.defaults = {
    version: 'ics20-1',
};
function channel(flags, logger) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        const home = resolve_home_option_1.resolveHomeOption({ homeFlag: flags.home });
        const app = load_and_validate_app_1.loadAndValidateApp(home);
        if (!app) {
            throw new Error(`${constants_1.appFile} not found at ${home}`);
        }
        const keyFile = resolve_key_file_option_1.resolveKeyFileOption({ keyFileFlag: flags.keyFile, app });
        const mnemonic = yield resolve_mnemonic_option_1.resolveMnemonicOption({
            interactiveFlag: flags.interactive,
            mnemonicFlag: flags.mnemonic,
            keyFile,
            app,
        });
        const src = resolve_option_1.resolveOption('src', { required: true })(app.src);
        const dest = resolve_option_1.resolveOption('dest', { required: true })(app.dest);
        const srcConnection = resolve_option_1.resolveOption('srcConnection', { required: true })(flags.srcConnection, app.srcConnection, process.env.RELAYER_SRC_CONNECTION);
        const destConnection = resolve_option_1.resolveOption('destConnection', { required: true })(flags.destConnection, app.destConnection, process.env.RELAYER_DEST_CONNECTION);
        const srcPort = resolve_option_1.resolveOption('srcPort', { required: true })(flags.srcPort, process.env.RELAYER_SRC_PORT);
        const destPort = resolve_option_1.resolveOption('destPort', { required: true })(flags.destPort, process.env.RELAYER_DEST_PORT);
        const version = (_a = resolve_option_1.resolveOption('version')(flags.version, process.env.RELAYER_VERSION)) !== null && _a !== void 0 ? _a : exports.defaults.version;
        const options = {
            home,
            mnemonic,
            src,
            dest,
            srcConnection,
            destConnection,
            srcPort,
            destPort,
            version,
            ordered: flags.ordered,
        };
        yield run(options, logger);
    });
}
exports.channel = channel;
function run(options, logger) {
    return __awaiter(this, void 0, void 0, function* () {
        const registryFilePath = path_1.default.join(options.home, constants_1.registryFile);
        const registry = load_and_validate_registry_1.loadAndValidateRegistry(registryFilePath);
        const srcChain = registry.chains[options.src];
        if (!srcChain) {
            throw new Error(`src channel "${options.src}" not found in registry`);
        }
        const destChain = registry.chains[options.dest];
        if (!destChain) {
            throw new Error(`dest channel "${options.dest}" not found in registry`);
        }
        const nodeA = yield signing_client_1.signingClient(srcChain, options.mnemonic, logger.child({ label: srcChain.chain_id }));
        const nodeB = yield signing_client_1.signingClient(destChain, options.mnemonic, logger.child({ label: destChain.chain_id }));
        const link = yield link_1.Link.createWithExistingConnections(nodeA, nodeB, options.srcConnection, options.destConnection, logger);
        const ordering = options.ordered
            ? channel_1.Order.ORDER_ORDERED
            : channel_1.Order.ORDER_UNORDERED;
        const channel = yield link.createChannel('A', options.srcPort, options.destPort, ordering, options.version);
        const output = [
            'Created channel:',
            ...indent_1.indent([
                `${srcChain.chain_id}: ${channel.src.portId}/${channel.src.channelId} (${link.endA.connectionID})`,
                `${destChain.chain_id}: ${channel.dest.portId}/${channel.dest.channelId} (${link.endB.connectionID})`,
            ]),
        ].join(os_1.default.EOL);
        console.log(output);
    });
}
exports.run = run;
