// @confio/relayer/build/binary/ibc-setup/commands/connect.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = exports.connect = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const link_1 = require("../../../lib/link");
const constants_1 = require("../../constants");
const load_and_validate_app_1 = require("../../utils/load-and-validate-app");
const load_and_validate_registry_1 = require("../../utils/load-and-validate-registry");
const resolve_option_1 = require("../../utils/options/resolve-option");
const resolve_home_option_1 = require("../../utils/options/shared/resolve-home-option");
const resolve_key_file_option_1 = require("../../utils/options/shared/resolve-key-file-option");
const resolve_mnemonic_option_1 = require("../../utils/options/shared/resolve-mnemonic-option");
const signing_client_1 = require("../../utils/signing-client");
function connect(flags, logger) {
    return __awaiter(this, void 0, void 0, function* () {
        const home = resolve_home_option_1.resolveHomeOption({ homeFlag: flags.home });
        const app = load_and_validate_app_1.loadAndValidateApp(home);
        if (!app) {
            throw new Error(`${constants_1.appFile} not found at ${home}`);
        }
        const keyFile = resolve_key_file_option_1.resolveKeyFileOption({ keyFileFlag: flags.keyFile, app });
        const mnemonic = yield resolve_mnemonic_option_1.resolveMnemonicOption({
            interactiveFlag: flags.interactive,
            mnemonicFlag: flags.mnemonic,
            keyFile,
            app,
        });
        const src = resolve_option_1.resolveOption('src', { required: true })(app.src);
        const dest = resolve_option_1.resolveOption('dest', { required: true })(app.dest);
        const srcTrust = resolve_option_1.resolveOption('srcTrust', { integer: true })(flags.srcTrust, process.env.RELAYER_SRC_TRUST);
        const destTrust = resolve_option_1.resolveOption('destTrust', { integer: true })(flags.destTrust, process.env.RELAYER_DEST_TRUST);
        const options = {
            home,
            mnemonic,
            src,
            dest,
            srcTrust,
            destTrust,
        };
        yield run(options, app, logger);
    });
}
exports.connect = connect;
function run(options, app, logger) {
    return __awaiter(this, void 0, void 0, function* () {
        const registryFilePath = path_1.default.join(options.home, constants_1.registryFile);
        const registry = load_and_validate_registry_1.loadAndValidateRegistry(registryFilePath);
        const srcChain = registry.chains[options.src];
        if (!srcChain) {
            throw new Error(`src channel  "${options.src}" not found in registry`);
        }
        const destChain = registry.chains[options.dest];
        if (!destChain) {
            throw new Error(`dest channel  "${options.dest}" not found in registry`);
        }
        const nodeA = yield signing_client_1.signingClient(srcChain, options.mnemonic, logger);
        const nodeB = yield signing_client_1.signingClient(destChain, options.mnemonic, logger);
        const link = yield link_1.Link.createWithNewConnections(nodeA, nodeB, logger, options.srcTrust, options.destTrust);
        const appYaml = js_yaml_1.default.dump(Object.assign(Object.assign({}, app), { srcConnection: link.endA.connectionID, destConnection: link.endB.connectionID }), {
            lineWidth: 1000,
        });
        fs_1.default.writeFileSync(path_1.default.join(options.home, constants_1.appFile), appYaml, {
            encoding: 'utf-8',
        });
        console.log(`Created connections ${link.endA.connectionID} (${link.endA.clientID}) <=> ${link.endB.connectionID} (${link.endB.clientID})`);
    });
}
exports.run = run;
