// @confio/relayer/build/binary/ibc-setup/commands/ics20.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = exports.ics20 = exports.defaults = void 0;
const fs_1 = __importDefault(require("fs"));
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const channel_1 = require("../../../codec/ibc/core/channel/v1/channel");
const link_1 = require("../../../lib/link");
const constants_1 = require("../../constants");
const indent_1 = require("../../utils/indent");
const load_and_validate_app_1 = require("../../utils/load-and-validate-app");
const load_and_validate_registry_1 = require("../../utils/load-and-validate-registry");
const resolve_option_1 = require("../../utils/options/resolve-option");
const resolve_home_option_1 = require("../../utils/options/shared/resolve-home-option");
const resolve_key_file_option_1 = require("../../utils/options/shared/resolve-key-file-option");
const resolve_mnemonic_option_1 = require("../../utils/options/shared/resolve-mnemonic-option");
const signing_client_1 = require("../../utils/signing-client");
exports.defaults = {
    port: 'transfer',
};
function resolveConnections({ srcConnection, destConnection, }) {
    if (!srcConnection && destConnection) {
        throw new Error(`You have defined "destConnection" but no "srcConnection". Both "srcConnection" and "destConnection" must be present.`);
    }
    if (srcConnection && !destConnection) {
        throw new Error(`You have defined "srcConnection" but no "destConnection". Both "srcConnection" and "destConnection" must be present.`);
    }
    if (srcConnection && destConnection) {
        return {
            src: srcConnection,
            dest: destConnection,
        };
    }
    return null;
}
function ics20(flags, logger) {
    return __awaiter(this, void 0, void 0, function* () {
        const home = resolve_home_option_1.resolveHomeOption({ homeFlag: flags.home });
        const app = load_and_validate_app_1.loadAndValidateApp(home);
        if (!app) {
            throw new Error(`${constants_1.appFile} not found at ${home}`);
        }
        const keyFile = resolve_key_file_option_1.resolveKeyFileOption({ keyFileFlag: flags.keyFile, app });
        const mnemonic = yield resolve_mnemonic_option_1.resolveMnemonicOption({
            interactiveFlag: flags.interactive,
            mnemonicFlag: flags.mnemonic,
            keyFile,
            app,
        });
        const src = resolve_option_1.resolveOption('src', { required: true })(app.src);
        const dest = resolve_option_1.resolveOption('dest', { required: true })(app.dest);
        // we apply default ports later, once we have the registry
        const srcPort = resolve_option_1.resolveOption('srcPort')(flags.srcPort, process.env.RELAYER_SRC_PORT);
        const destPort = resolve_option_1.resolveOption('destPort')(flags.destPort, process.env.RELAYER_DEST_PORT);
        const srcTrust = resolve_option_1.resolveOption('srcTrust', { integer: true })(flags.srcTrust, process.env.RELAYER_SRC_TRUST);
        const destTrust = resolve_option_1.resolveOption('destTrust', { integer: true })(flags.destTrust, process.env.RELAYER_DEST_TRUST);
        const connections = resolveConnections(app);
        yield run({
            src,
            dest,
            home,
            mnemonic,
            srcPort,
            destPort,
            connections,
            srcTrust,
            destTrust,
        }, app, logger);
    });
}
exports.ics20 = ics20;
function resolveLink(nodeA, nodeB, { connections, srcTrust, destTrust }, logger) {
    return __awaiter(this, void 0, void 0, function* () {
        if (connections) {
            const link = yield link_1.Link.createWithExistingConnections(nodeA, nodeB, connections.src, connections.dest, logger);
            console.log(`Used existing connections [${link.endA.chainId()}, ${link.endA.connectionID}, ${link.endA.clientID}] <=> [${link.endB.chainId()}, ${link.endB.connectionID}, ${link.endB.clientID}]`);
            return link;
        }
        const link = yield link_1.Link.createWithNewConnections(nodeA, nodeB, logger, srcTrust, destTrust);
        console.log(`Created connections [${link.endA.chainId()}, ${link.endA.connectionID}, ${link.endA.clientID}] <=> [${link.endB.chainId()}, ${link.endB.connectionID}, ${link.endB.clientID}]`);
        return link;
    });
}
function run(options, app, logger) {
    return __awaiter(this, void 0, void 0, function* () {
        const registryFilePath = path_1.default.join(options.home, constants_1.registryFile);
        const { chains } = load_and_validate_registry_1.loadAndValidateRegistry(registryFilePath);
        const srcChain = chains[options.src];
        if (!srcChain) {
            throw new Error('src chain not found in registry');
        }
        const destChain = chains[options.dest];
        if (!destChain) {
            throw new Error('dest chain not found in registry');
        }
        const ordering = channel_1.Order.ORDER_UNORDERED;
        const version = 'ics20-1';
        const nodeA = yield signing_client_1.signingClient(srcChain, options.mnemonic, logger);
        const nodeB = yield signing_client_1.signingClient(destChain, options.mnemonic, logger);
        const link = yield resolveLink(nodeA, nodeB, options, logger);
        const srcConnection = link.endA.connectionID;
        const destConnection = link.endB.connectionID;
        const appFilePath = path_1.default.join(options.home, constants_1.appFile);
        const appYaml = js_yaml_1.default.dump(Object.assign(Object.assign({}, app), { srcConnection,
            destConnection }), {
            lineWidth: 1000,
        });
        fs_1.default.writeFileSync(appFilePath, appYaml, { encoding: 'utf-8' });
        // provide default port, either from registry or global default
        const srcPort = resolve_option_1.resolveOption('src-port', { required: true })(options.srcPort, srcChain.ics20_port, exports.defaults.port);
        const destPort = resolve_option_1.resolveOption('dest-port', { required: true })(options.destPort, destChain.ics20_port, exports.defaults.port);
        const channel = yield link.createChannel('A', srcPort, destPort, ordering, version);
        const output = [
            'Created channel:',
            ...indent_1.indent([
                `${srcChain.chain_id}: ${channel.src.portId}/${channel.src.channelId} (${link.endA.connectionID})`,
                `${destChain.chain_id}: ${channel.dest.portId}/${channel.dest.channelId} (${link.endB.connectionID})`,
            ]),
        ].join(os_1.default.EOL);
        console.log(output);
    });
}
exports.run = run;
